index.rst

```diff
@@ -1,61 +1,504 @@
 
-   Feel free to delete this instructional comment.
+- The ``QuantumGraph`` itself is a complete description of the processing that can be used to reproduce a ``RUN`` exactly (aside from transient hardware/environment problems, of course), on any system with access to the same datasets and sufficiently similar hardware and software (with the latter verifiable via the per-``RUN`` software versions dataset).
 
-:tocdepth: 1
+The primary limitation in the current system is that the ``QuantumGraph`` is at most saved to a file outside the data repository (at the user's discretion), and hence much of the provenance we produce is often discarded.
+It also only captures *predicted* inputs and outputs, which are in general a superset of the *actual* inputs consumed and outputs produced by quanta; this is sufficient for reproduction when starting everything from scratch, but harder to make use of when reproducing subsets or starting from a partial run.
+Saving a more complete, as-executed ``QuantumGraph`` in the data repository has always been part of the high-level middleware design, and a more detailed design is the subject of the next section.
 
-.. Please do not modify tocdepth; will be fixed when a new Sphinx theme is shipped.
+Before moving on, it is worth pointing out a few issues with the provenance that we do record:
 
-.. sectnum::
+- The software versions we record may not extend far enough into OS- or container-level software to guarantee reproducibility (but by using ``conda`` even for compilers, the dependence of our code's behavior on OS- and container-level software should be minimized).
 
-.. TODO: Delete the note below before merging new content to the master branch.
+- Because we do not assume that the complete set of relevant software packages can be enumerated, and instead rely (at least partially) on what Python packages are imported, running additional processing that brings in new imports against an existing output ``RUN`` collection requires us to update the versions dataset in place - an operation we otherwise prohibit in ``Butler`` (datasets are considered atomic and immutable, at least for writes) and simulate by deleting the old one and writing a new one.
+  This is both fragile and incompatible with our plans to distribute Data Release processing across multiple Data Facilities, unless we just prohibit that kind of ``RUN`` collection extension or do enumerate all relevant software somehow up front.
+
+- The approach of using datasets to record arbitrary provenance for higher-level tooling is currently extensible only to the extent that those datasets can use existing storage classes.
+  New storage classes must currently be registered by editing configuration in the data repository or ``daf_butler`` manually.
+  This is a problem for much more than just provenance, however (it also limits pipeline extensibility), and is something we hope to resolve.
```

Comment:
Butler does allow you to define your own storage classes anywhere and set an environment variable to tell butler where to find the new definitions. We've never really used that feature though.

Reply:
Added some text to mention this.